linera-storage/src/db_storage.rs

```diff
@@ -604,36 +616,42 @@ where
     }
 
     async fn contains_certificate(&self, hash: CryptoHash) -> Result<bool, ViewError> {
-        let cert_key = bcs::to_bytes(&BaseKey::Certificate(hash))?;
-        let value_key = bcs::to_bytes(&BaseKey::CertificateValue(hash))?;
-        let keys = vec![cert_key, value_key];
+        let keys = Self::construct_query(&[hash])?;
         let results = self.store.contains_keys(keys).await?;
         #[cfg(with_metrics)]
         CONTAINS_CERTIFICATE_COUNTER.with_label_values(&[]).inc();
         Ok(results[0] && results[1])
     }
 
     async fn read_certificate(&self, hash: CryptoHash) -> Result<Certificate, ViewError> {
-        let cert_key = bcs::to_bytes(&BaseKey::Certificate(hash))?;
-        let value_key = bcs::to_bytes(&BaseKey::CertificateValue(hash))?;
-        let keys = vec![cert_key, value_key];
+        let keys = Self::construct_query(&[hash])?;
         let values = self.store.read_multi_values_bytes(keys).await;
         if values.is_ok() {
             #[cfg(with_metrics)]
             READ_CERTIFICATE_COUNTER.with_label_values(&[]).inc();
         }
         let values = values?;
-        let cert_bytes = values[0]
-            .as_ref()
-            .ok_or_else(|| ViewError::not_found("certificate bytes for hash", hash))?;
-        let value_bytes = values[1]
-            .as_ref()
-            .ok_or_else(|| ViewError::not_found("value bytes for hash", hash))?;
-        let cert = bcs::from_bytes::<LiteCertificate>(cert_bytes)?;
-        let value = bcs::from_bytes::<CertificateValue>(value_bytes)?;
-        Ok(cert
-            .with_value(value.with_hash_unchecked(hash))
-            .ok_or(ViewError::InconsistentEntries)?)
+        Self::deserialize_certificate(&values, hash)
+    }
+
+    async fn read_certificates<I: IntoIterator<Item = CryptoHash> + Send>(
+        &self,
+        hashes: I,
+    ) -> Result<Vec<Certificate>, ViewError> {
+        let hashes = hashes.into_iter().collect::<Vec<_>>();
+        let keys = Self::construct_query(&hashes)?;
```

Comment:
Can we pass `hashes` right away to `construct_query`, without `into_iter().collect()` ? Right now the caller of `read_certificates` most likely passes some kind of `vec!` already and here we'd call `into_iter().to_vec()` on it (so an unnecessary transformation 🤔 ).

Maybe changing the argument's type to `impl IntoIterator<Item=CryptoHash>` is needed?

Reply:
The answer is no.
The reason is that `IntoIterator` implements the `into_iter` which consumes the object. So, we can consumes only once.
The transformation into hashes is a little bit suboptimal I agree. But I think it is fine. We could replace the argument `impl IntroIterator<Item=CryptoHash>` to something else but maybe this is something for later.